Sources/AWSLambdaRuntimeCore/Terminator.swift

```diff
@@ -0,0 +1,133 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftAWSLambdaRuntime open source project
+//
+// Copyright (c) 2022 Apple Inc. and the SwiftAWSLambdaRuntime project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftAWSLambdaRuntime project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIOConcurrencyHelpers
+import NIOCore
+
+extension Lambda {
+    /// Lambda terminator.
+    /// Utility to manage the lambda shutdown sequence.
+    public final class Terminator {
+        public typealias Handler = (EventLoop) -> EventLoopFuture<Void>
+        public typealias RegistrationKey = String
+
+        private var storage: Storage
+
+        public init() {
+            self.storage = Storage()
+        }
+
+        /// Register a shutdown handler with the terminator
+        ///
+        /// - parameters:
+        ///     - name: Display name for logging purposes
+        ///     - handler: The shutdown handler to call when terminating the Lambda.
+        ///             Shutdown handlers are called in the reverse order of being registered.
+        ///
+        /// - Returns: A `RegistrationKey` that can be used to de-register the handler when its no longer needed.
+        @discardableResult
+        public func register(name: String, handler: @escaping Handler) -> RegistrationKey {
```

Comment:
As you know, I'm not a big fan of the typealiases for closures, since they destroy code completion.